File: app/src/main/java/com/katkov/lolachievements/data/mappers/Mapper.java
package com.katkov.lolachievements.data.mappers;

import java.util.ArrayList;
import java.util.List;

public abstract class Mapper<S, D> {

    public abstract D map(S source);

    public final List<D> mapList(List<S> sourceItems) {
        List<D> result = new ArrayList<>();

        for (final S item : sourceItems) {
            result.add(map(item));
        }

        return result;
    }
}

File: app/src/main/java/com/katkov/lolachievements/data/mappers/Mapper.kt
package com.katkov.lolachievements.data.mappers

import java.util.*

abstract class Mapper<S, D> {

    abstract fun map(source: S): D

    fun mapList(sourceItems: List<S>): List<D> {
        val result = ArrayList<D>()

        for (item in sourceItems) {
            result.add(map(item))
        }

        return result
    }
}